### PR TITLE
MNT-20764: Searches fail for users who are members of groups where the authorityName contains double quotes

### PR DESCRIPTION
**Pull request by amohammedalfresco**
- Guard QName string against a list of invalid characters `{'/', '\\', '\n', '\r', '"'}`

### CODE REVIEW

File: src/main/java/org/alfresco/service/namespace/QName.java [hunk 1]
```diff
@@ -215,7 +219,28 @@ public static String createValidLocalName(String name)
...
     }
-    
+
+    /**
+     * Check for special characters in name
+     *
+     * @param name  name to create valid local name from
+     * @throws IllegalArgumentException
+     */
+    public static void checkName(String name)
```
Comment (killerboot) [inline src/main/java/org/alfresco/service/namespace/QName.java]: I would like to propose to do some extra for this method.
Either rename it to something like "checkLocalNameForInvalidCharacters" as there are other checks in the code like the check for null in "createQName" methods or move the other checks  ("(localName == null || localName.length() == 0)) into this method and rename it to "checkLocalName"
I would prefer the second.

The other note is that if you have an exception thrown in this method, the signature should mention that. You may have noticed that the checks for non present local name throw InvalidQNameException whereas your method throw a different IllegalArgumentException. I would make this exception the same.

**Review by killerboot — CHANGES_REQUESTED** [review]

File: src/main/java/org/alfresco/service/namespace/QName.java (hunk 1 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/service/namespace/QName.java]: Sure.  Thanks for the great pointers.

File: src/main/java/org/alfresco/service/namespace/QName.java [hunk 2]
```diff
@@ -203,19 +190,42 @@ public static QName createQNameWithValidLocalName(final String namespaceURI, Str
...
-    
+
+    /**
+     * Condition to check string name provided for QName
+     *
+     * @param name  name to create valid local name from
+     * @throws InvalidQNameException
+     */
+    public static void checkLocalName(String name)
+        throws IllegalArgumentException
```
Comment (killerboot) [inline src/main/java/org/alfresco/service/namespace/QName.java]: The exception you are throwing here is InvalidQNameException, not IllegalArgumentException.

File: src/main/java/org/alfresco/service/namespace/QName.java [hunk 3]
```diff
@@ -147,11 +142,6 @@ public static QName createQName(String qname, NamespacePrefixResolver prefixReso
     public static QName createQName(String qname)
         throws InvalidQNameException
     {
-        if (qname == null || qname.length() == 0)
```
Comment (killerboot) [inline src/main/java/org/alfresco/service/namespace/QName.java]: If you just remove this check, the code further on will throw NullPointerExceptions, for example line 167 `namespaceEnd = qname.indexOf(NAMESPACE_END, namespaceBegin + 1);`

**Review by killerboot — CHANGES_REQUESTED** [review]

File: src/main/java/org/alfresco/service/namespace/QName.java (hunk 2 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/service/namespace/QName.java]: Thanks for the pointer, I totally forgot to replace the signature with new one.

File: src/main/java/org/alfresco/service/namespace/QName.java (hunk 3 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/service/namespace/QName.java]: I just reverted that null check.  Thanks.

File: src/main/java/org/alfresco/service/namespace/QName.java [hunk 4]
```diff
@@ -203,19 +195,42 @@ public static QName createQNameWithValidLocalName(final String namespaceURI, Str
...
     }
-    
+
+    /**
+     * Condition to check string name provided for QName
+     *
+     * @param name  name to create valid local name from
+     * @throws InvalidQNameException
+     */
+    public static void checkLocalName(String name)
```
Comment (killerboot) [inline src/main/java/org/alfresco/service/namespace/QName.java]: It may be better to make this method package private. I'm not sure it can be used anywhere outside this class.

**Review by killerboot — CHANGES_REQUESTED** [review]
Please add unit tests for all public methods that now have this check to make sure this doesn't regress in future.

File: src/main/java/org/alfresco/service/namespace/QName.java (hunk 4 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/service/namespace/QName.java]: Method converted to `private`.

**Comment by amohammedalfresco** [comment]
> Please add unit tests for all public methods that now have this check to make sure this doesn't regress in future.

I've added more tests to check other methods where `checkLocalName()` is used.

**Review by killerboot — CHANGES_REQUESTED** [review]
The tests look good. Can you please add similar tests for `null` and empty strings for the same methods?

**Comment by amohammedalfresco** [comment]
> The tests look good. Can you please add similar tests for `null` and empty strings for the same methods?

Sure.

File: src/main/java/org/alfresco/service/namespace/QName.java [hunk 5]
```diff
@@ -203,19 +195,42 @@ public static QName createQNameWithValidLocalName(final String namespaceURI, Str
...
+            throw new InvalidQNameException("A QName must consist of a local name and cannot be null or empty");
+        }
+
+        if (name != null)
+        {
+            for (char illegalCharacter : illegalCharacters)
+            {
+                if (name.indexOf(illegalCharacter) != -1)
+                {
+                    throw new InvalidQNameException("name contains characters that are not permitted: "+name.charAt(name.indexOf(illegalCharacter)));
```
Comment (killerboot) [inline src/main/java/org/alfresco/service/namespace/QName.java]: The message in this exception should start from uppercase letter, also I think it should specify that it is a "local name".
I noticed that you are checking for the exact message in tests. The checks are duplicated and will have to be fixed everywhere is somebody changes the message like in this case. It is better to avoid the checks for the exact message unless it is absolutely necessary to ensure the right contents are present. I would recommend to just check for the exception of a particular class.

File: src/test/java/org/alfresco/service/namespace/QNameTest.java [hunk 6]
```diff
@@ -124,6 +124,211 @@ public void testInvalidQName() throws Exception
...
+                QName.createQName(NamespaceService.ALFRESCO_PREFIX, localName);
+            }
+            catch (InvalidQNameException e)
+            {
+                assertEquals(e.getMessage(), "name contains characters that are not permitted: "+localName.charAt(localName.indexOf(illegalCharacter)));
+            }
+        }
+    }
+
+    public void testCreateQnameWithNoPrefixWithNullAndEmptyCharacters()
```
Comment (killerboot) [inline src/test/java/org/alfresco/service/namespace/QNameTest.java]: There is a lot of duplication in the tests. This should be avoided.
A cleaner way to test would be something like this:
```
    @Test(expected = InvalidQNameException.class)
    public void testCreateQnameWithNoPrefixWithNull()
    {
        String nullLocalName = null;
        QName.createQName(NamespaceService.ALFRESCO_PREFIX, nullLocalName);
    }
```

**Review by killerboot — CHANGES_REQUESTED** [review]

File: src/test/java/org/alfresco/service/namespace/QNameTest.java (hunk 6 above)
Comment (amohammedalfresco) [reply src/test/java/org/alfresco/service/namespace/QNameTest.java]: I noticed the existing tests are written using JUnit3 and the use of annotation `@Test(expected = InvalidQNameException.class)` is not working as expected :(

I've read something related to this:
> JUnit4 style tests should not extend junit.framework.TestCase directly or indirectly. If you do, you often get confusing behavior (fields annotated by @Rule ignored, test failures because the base class methods are not being called by JUnit, etc). The tests may be run as JUnit4-style or JUnit3-style depending bon how they are run.
> 
> Your options are either to stick with JUnit3 or convince the maintainers of your base class to create classes that support JUnit4 (Rules, Runners, etc).

Comment (amohammedalfresco) [reply src/test/java/org/alfresco/service/namespace/QNameTest.java]: However I will try to reduce the noise by using a simpel `try/catch` method to only capture the exception (like other tests).

File: src/main/java/org/alfresco/service/namespace/QName.java (hunk 5 above)
Comment (amohammedalfresco) [reply src/main/java/org/alfresco/service/namespace/QName.java]: Sure. Done!

File: src/test/java/org/alfresco/service/namespace/QNameTest.java (hunk 6 above)
Comment (amohammedalfresco) [reply src/test/java/org/alfresco/service/namespace/QNameTest.java]: After removing the `extends TestCase` and removing the public constructor method that was causing problems in using `@Test` annotation, the Tests are now refactored to just catch the expected exception.

**Review by killerboot — APPROVED** [review]
